ThunkLibs/libOpenCL/libOpenCL_Common.h

```diff
@@ -0,0 +1,229 @@
+        Type == AsyncCallbackType::TYPE_CLCREATECONTEXTFROMTYPE ||
+        Type == AsyncCallbackType::TYPE_CLSETCONTEXTDESTRUCTORCALLBACK ||
+        Type == AsyncCallbackType::TYPE_CLSETMEMOBJECTDESTRUCTORCALLBACK ||
+        Type == AsyncCallbackType::TYPE_CLSETPROGRAMRELEASECALLBACK ||
+        Type == AsyncCallbackType::TYPE_CLSETEVENTCALLBACK ||
+        Type == AsyncCallbackType::TYPE_CLBUILDPROGRAM ||
+        Type == AsyncCallbackType::TYPE_CLCOMPILEPROGRAM ||
+        Type == AsyncCallbackType::TYPE_CLLINKPROGRAM ||
+        Type == AsyncCallbackType::TYPE_CLENQUEUESVMFREE
+        ,
+        "Mismatch type for handling clProgram");
+
+      FEX::CL::Type_SingleUserData Data {
+        .Header {
+          .Type = Type,
+          .cb = a_0,
+        },
+        .user_data = a_1,
+      };
+
+      return Data;
+    };
+  };
+
+  struct Type_SingleUserDataCopy {
+    CallbackData::_Header Header;
+
+    // User provided data that we need to keep around
+    void *user_data;
+    size_t size;
+    ~Type_SingleUserDataCopy() {
+      free(user_data);
+    }
+
+    template <AsyncCallbackType Type>
+    static std::unique_ptr<Type_SingleUserDataCopy> Create(decltype(CallbackData::_Header::cb) a_0, void * a_1, size_t size) {
+      static_assert(Type == AsyncCallbackType::TYPE_CLENQUEUENATIVEKERNEL,
+        "Mismatch type for handling clProgram");
+
+      void *Memory = malloc(size);
+      memcpy(Memory, a_1, size);
+      return std::make_unique<FEX::CL::Type_SingleUserDataCopy>(FEX::CL::Type_SingleUserDataCopy{
+        .Header {
+          .Type = Type,
+          .cb = a_0,
+        },
+        .user_data = Memory,
+        .size = size,
+      });
+    };
+  };
+
+  union GuestCallbackData {
+    CallbackData::_Header Header;
+
+    struct Type_clCreateContext {
+      CallbackData::_Header Header;
+      const char *a_0;
+      const void *a_1;
+      size_t a_2;
+      void *a_3;
+
+      static Type_clCreateContext Create(Type_SingleUserData* OriginalData, const char *a_0, const void *a_1, size_t a_2) {
+        Type_clCreateContext Data;
+        Data.Header.Type = OriginalData->Header.Type;
+        Data.Header.cb = OriginalData->Header.cb;
+        Data.a_0 = a_0;
+        Data.a_1 = a_1;
+        Data.a_2 = a_2;
+        Data.a_3 = OriginalData->user_data;
+        return Data;
+      }
+    } clCreateContext;
+
+    struct Type_clSetContextDestructorCallback {
+      CallbackData::_Header Header;
+
+      cl_context a_0;
+      void *a_1;
+
+      static Type_clSetContextDestructorCallback Create(Type_SingleUserData* OriginalData, cl_context a_0) {
+        Type_clSetContextDestructorCallback Data;
+        Data.Header.Type = OriginalData->Header.Type;
+        Data.Header.cb = OriginalData->Header.cb;
+        Data.a_0 = a_0;
+        Data.a_1 = OriginalData->user_data;
+        return Data;
+      }
+    } clSetContextDestructorCallback;
+
+    struct Type_clProgram {
```

Comment:
This seems to duplicate a lot of functionality provided by `PackedArguments`. Couldn't most of the `Type_` definitions be replaced with a template struct?

Reply:
Maybe? Sounds like it would make the interface more complex to read. Only three have the same arrangement of arguments where a_0 changes between `cl_program`, `cl_mem`, and `cl_context`. The other four all have different arrangements of arguments and of course POD is still a requirement.

Reply:
Would something like this work? https://godbolt.org/z/dTKEooMdo

Generally I'd like to avoid playing manual code generator in thunking code, since it makes structural changes harder to pull off in the future.

Reply:
You're right, that does work. Changed the code over to use this.